### ISSUE TITLE
Marker Clustering

### ISSUE THREAD

**Issue by getmetorajesh** [Yermo/nativescript-mapbox#33]
Any plans on implementing the cluster functionality in [mapbox](https://www.mapbox.com/mapbox-gl-js/example/cluster/)

**Comment by EddyVerbruggen** [comment]
Hi, I'd love to but you're pointing to the JS version whereas this plugin uses the native SDK's. If you can find the clustering feature for native I'd love to hear! Otherwise please ask in the Mapbox GitHub repo, I'm sure they've heard the request before.

**Comment by EddyVerbruggen** [comment]
Clustering has recently been added to the native SDK (on Android in v 4.2.0) so it's time to look at this again!

**Comment by EddyVerbruggen** [comment]
2.5.0 has marker clustering on Android, but I couldn't get unclustered markers to show up so it's an undocumented feature for now.

**Comment by rohanrichards** [comment]
Hey @EddyVerbruggen check out [this example](https://www.mapbox.com/mapbox-gl-js/example/cluster/). I know it's for the gl.js library, however it does show that they show unclustered markers by adding a second layer. 

They use the Filter property to hide it when the point has the point count layer visible (meaning it's been clustered).

Example: 
```
    map.addLayer({
        id: "unclustered-point",
        type: "circle",
        source: "earthquakes",
        filter: ["!has", "point_count"],
        paint: {
            "circle-color": "#11b4da",
            "circle-radius": 4,
            "circle-stroke-width": 1,
            "circle-stroke-color": "#fff"
        }
    });
```

This is the same approach I used in a project using ngx-mapbox (which is built on mapbox-gl) and it worked well in the browser, perhaps it could work with your native implementation?

**Comment by keerl** [comment]
It looks like clustering has been added to iOS SDK. Is this a feature that could be looked into again? And possibly documented for Android as well?

[https://www.mapbox.com/ios-sdk/maps/examples/clustering/](https://www.mapbox.com/ios-sdk/maps/examples/clustering/)

**Comment by paul-muckypuddle** [comment]
Agreed, would love to see this in the library!

**Comment by Codex-** [comment]
Very keen to see this added! Some solid examples around now for both iOS and Android.

Edit: Oh, as said previously, it's already there: https://github.com/EddyVerbruggen/nativescript-mapbox/blob/master/src/mapbox.common.ts#L187

**Comment by Tyler-V** [comment]
Clustering would be a great addition, from what I can see there aren't any NativeScript plugins that have map clustering / heatmap options for both Android and iOS.

**Comment by Codex-** [comment]
@Tyler-V You can use this [currently](https://github.com/EddyVerbruggen/nativescript-mapbox/blob/master/src/mapbox.common.ts#L187) ;)

I tested with Android and it seems to work, whether or not it works well in the current version is another question, I've been meaning to look into it more :(

**Comment by Tyler-V** [comment]
Yes they both implement mapbox.common.ts but only Android currently implements that method, iOS is [not available yet](https://github.com/EddyVerbruggen/nativescript-mapbox/blob/master/src/mapbox.ios.ts#L1035).

I was looking to do a mapping project in NativeScript but it appears that there just aren't any mature libraries that meet all of my requirements (heatmap, clustering, offline)

**Comment by Reached** [comment]
+1 for clustering (IOS/Android) :)

Also I am willing to donate/pay for someone to do this. Unfortunately I am unable to contribute myself on this part :/

**Comment by Sauvetonbio** [comment]
Hello, any updates to implement clustering on iOS+Android ?
We absolutly need it for our project, as we are already using it in our production website...
Do you know if it's technically possible to implement it our-self, or something could stop us ? 

Example iOS: https://docs.mapbox.com/ios/maps/examples/clustering/
Example Android: https://docs.mapbox.com/android/maps/examples/circle-layer-clustering/

Thank you

**Comment by mohammadrafigh** [comment]
I'm going to make a PR both for Android and iOS.

**Comment by sguimont** [comment]
What is missing for the PR to be integrated?

**Comment by Yermo** [comment]
@sguimont This PR is old from before I took the project over. It needs to be updated to match the v5 changes. I haven't had a chance to look at it yet.

**Comment by mohammadrafigh** [comment]
I will update PR asap.

**Comment by sguimont** [comment]
@mohammadrafigh You will make my day!  I'm happy to hear someone will continue the work of this feature.

**Comment by Sauvetonbio** [comment]
I’m still waiting since October too 🙃 thank you !!

**Comment by Yermo** [comment]
@mohammadrafigh If at all possible, can you make your API match how the equivalent would be done in the Mapbox GL JS API?

What I want to do with this library is make it call compatible with Mapbox GL JS to make porting between web and native. It makes a very compelling use case for adopting NativeScript and this plugin as a solution.

**Comment by mohammadrafigh** [comment]
@Yermo Actually I'm using nativescript-mapbox and Mapbox GL JS in a shared project (schematics). So using the same API is preferred for me too. For sure I will try to do my best but for now I need to find a workaround for #379 that I faced today when I started to do some investigations. please let me know if you have thoughts about it.

**Comment by b44rd** [comment]
@mohammadrafigh Did you find a workaround for #379? Would be very nice to see your pull request merged at some point. Let me know if there´s something I can help with in order to make it happen.

**Comment by Codex-** [comment]
@b44rd the readme advises that the project has now moved: https://github.com/nativescript-community/ui-mapbox
